types/wlr_xdg_foreign_v2.c

```diff
@@ -0,0 +1,411 @@
+#define _POSIX_C_SOURCE 200809L
+
+#include <assert.h>
+#include <string.h>
+#include <stdlib.h>
+#include <wlr/types/wlr_xdg_foreign_v2.h>
+#include <wlr/types/wlr_xdg_shell.h>
+#include <wlr/util/log.h>
+#include "util/signal.h"
+#include "xdg-foreign-unstable-v2-protocol.h"
+
+#define FOREIGN_V2_VERSION 1
+
+static const struct zxdg_exported_v2_interface xdg_exported_impl;
+static const struct zxdg_imported_v2_interface xdg_imported_impl;
+static const struct zxdg_exporter_v2_interface xdg_exporter_impl;
+static const struct zxdg_importer_v2_interface xdg_importer_impl;
+
+static struct wlr_xdg_imported_v2 *xdg_imported_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zxdg_imported_v2_interface,
+		&xdg_imported_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void xdg_imported_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static bool verify_is_toplevel(struct wl_resource *client_resource,
+		struct wlr_surface *surface) {
+	if (wlr_surface_is_xdg_surface(surface)) {
+		struct wlr_xdg_surface *xdg_surface =
+			wlr_xdg_surface_from_wlr_surface(surface);
+		if (xdg_surface->role != WLR_XDG_SURFACE_ROLE_TOPLEVEL) {
+			wl_resource_post_error(client_resource, -1,
+					"surface must be an xdg_toplevel");
+			return false;
+		}
+	} else {
+		wl_resource_post_error(client_resource, -1,
+				"surface must be an xdg_surface");
+		return false;
+	}
+	return true;
+}
+
+static void destroy_imported_child(struct wlr_xdg_imported_child_v2 *child) {
+	wl_list_remove(&child->xdg_toplevel_set_parent.link);
+	wl_list_remove(&child->xdg_surface_unmap.link);
+	wl_list_remove(&child->link);
+	free(child);
+}
+
+static void handle_child_xdg_surface_unmap(
+		struct wl_listener *listener, void *data) {
+	struct wlr_xdg_imported_child_v2 *child =
+		wl_container_of(listener, child, xdg_surface_unmap);
+	destroy_imported_child(child);
+}
+
+static void handle_xdg_toplevel_set_parent(
+		struct wl_listener *listener, void *data) {
+	struct wlr_xdg_imported_child_v2 *child =
+		wl_container_of(listener, child, xdg_toplevel_set_parent);
+	destroy_imported_child(child);
+}
+
+static void xdg_imported_handle_set_parent_of(struct wl_client *client,
+		struct wl_resource *resource,
+		struct wl_resource *child_resource) {
+	struct wlr_xdg_imported_v2 *imported =
+		xdg_imported_from_resource(resource);
+	if (imported->exported == NULL) {
+		return;
+	}
+	struct wlr_surface *wlr_surface = imported->exported->surface;
+	struct wlr_surface *wlr_surface_child =
+		wlr_surface_from_resource(child_resource);
+
+	if (!verify_is_toplevel(resource, wlr_surface_child)) {
+		return;
+	}
+	if (wlr_surface_is_xdg_surface(wlr_surface)
+			!= wlr_surface_is_xdg_surface(wlr_surface_child)) {
```

Comment:
Hm, `verify_is_toplevel` should take care of this too I think?

Reply:
Yes, it was a leftover from the time when there was both xdg6 and xdg-stable